rfcs/rfc-bld-20220623-1-build-failure-rca.md

```diff
@@ -0,0 +1,167 @@
+```
+  {
+    "_class" : "io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImpl",
+    "displayName" : "Android [release]",
+    "id" : "116",
+    "result" : "SUCCESS",
+    "type" : "PARALLEL",
+    "edges" : [
+      {
+        "_class" : "io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImpl$EdgeImpl",
+        "id" : "778",
+        "type" : "STAGE"
+      }
+    ]
+  },
+```
+2. **Setup Stage:** After the parallel stage is the setup stage, setup stage is to setup build node, pull source codes from Github and install build dependencies.
+```
+  {
+    "_class" : "io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImpl",
+    "displayName" : "Setup",
+    "id" : "778",
+    "result" : "SUCCESS",
+    "type" : "STAGE",
+    "edges" : [
+      {
+        "_class" : "io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImpl$EdgeImpl",
+        "id" : "1112",
+        "type" : "STAGE"
+      }
+    ]
+  },
+```
+3. **Build Stage:** Build stage is the stage runs the actual build task.
+```
+{
+    "_class" : "io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImpl",
+    "displayName" : "release",
+    "id" : "1112",
+    "result" : "SUCCESS",
+    "type" : "STAGE",
+    "edges" : [
+      {
+        "_class" : "io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImpl$EdgeImpl",
+        "id" : "1718",
+        "type" : "STAGE"
+      }
+    ]
+  },
+```
+The RCA dispatcher lambda function uses these pipeline stage data to build a relationship graph of build platform/configuration, and then it triggers the RCA worker lambda function to do build failure RCA for failed build platform/configuration.
+
+RCA worker lambda function sends a GET request
+```
+https://{JENKINS_URL}/blue/rest/organizations/jenkins/pipelines/{JOB_NAME}/runs/{BUILD_NUMBER}/nodes/{ID}/log
+```
+to get the build log of a certain build platform/configuration.
+#### Build Failure RCA Patterns
+RCA worker lambda function scans the build log and search for log messages that match build failure RCA patterns.
+
+Current build RCA patterns from https://github.com/aws-lumberyard/build-failure-rca/tree/main/build-failure-rca/rca-patterns can be reused.
+
+#### Workflow
+![workflow](rfc-bld-20220623-1-build-failure-rca/workflow.png)
+
+1. A change in build failure RCA repo triggers the build failure RCA CDK deployment pipeline.
+2. The build failure RCA CDK deployment pipeline deploys build failure RCA CDK stack to AWS accounts that are using build failure RCA system.
+3. Change in O3DE repo or O3DE fork repo triggers the AR or Periodic builds.
+4. After each AR or Periodic build completes, it publishes build data to post build SNS topic.
+5. Post build SNS topic triggers the RCA dispatcher lambda function.
+6. RCA dispatcher lambda function retrieves AR/Periodic build data and dispatch the actual RCA work for each build platform/configuration to separate RCA worker lambda functions. 
```

Comment:
SQS is also an option, I chose SNS because now we are already using a post build SNS to trigger MARS pipeline, creating AR Github issues and override PR status, so we don't need to maintain another SQS queue.

In terms of rate limiting, each AR/Periodic build only send one message to SNS and only triggers dispatcher lambda function once, the runs of worker lambda depends on the number of failed stages, and it won't exceed 30.
I checked https://docs.aws.amazon.com/sns/latest/dg/sns-message-delivery-retries.html#delivery-policy-maximum-receive-rate, and it seems that we are unlikely to have rate limiting issue unless we have hundreds of AR/Periodic pipeline sending SNS message simultaneously. We can create a improvement plan to use SQS or alternatives if we see  a rate limiting issue.

Reply:
Sounds good. If we end up adding a queue in front of the RCA lambda, we can still keep the original SNS for the intake from the AR. The SNS can then feed into SQS between that and the lambda. This would minimize any infra changes required to add it.

Reply:
Regarding the rate limiting it was more of a concern about the multiple AR runs completing instead of the number of stages. Is there a mechanism that we can add that will prevent an issue in the event that are hundreds of messages sent to the trigger SNS topic. Whether that's done through unintentional spamming or a config issue on our end.

Reply:
The nice thing about SQS is that it does FIFO and dedupe natively. There is a small risk that SNS could be sending out of order if a retry occurs for whatever period of time

Reply:
Added SQS queue in RFC for rate limiting